google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/BuildDeploymentSourceType.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 2016 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.cloud;
+
+import com.intellij.execution.BeforeRunTask;
+import com.intellij.execution.RunManagerEx;
+import com.intellij.execution.configurations.RunConfiguration;
+import com.intellij.execution.impl.ConfigurationSettingsEditorWrapper;
+import com.intellij.ide.DataManager;
+import com.intellij.openapi.actionSystem.DataContext;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.project.Project;
+import com.intellij.remoteServer.configuration.deployment.ModuleDeploymentSource;
+import com.intellij.remoteServer.impl.configuration.deployment.DeployToServerRunConfiguration;
+import com.intellij.remoteServer.impl.configuration.deployment.ModuleDeploymentSourceType;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.List;
+
+import javax.swing.JComponent;
+
+/**
+ * Base class for build system (e.g. maven or gradle) deployment source types.
+ */
+public abstract class BuildDeploymentSourceType extends ModuleDeploymentSourceType {
+
+  @Override
+  public void setBuildBeforeRunTask(
+      @NotNull RunConfiguration configuration,
+      @NotNull ModuleDeploymentSource source) {
+
+    Module module = source.getModule();
+
+    if(module == null) {
+      return;
+    }
+
+    setConfiguration(configuration);
+
+    RunManagerEx runManager = RunManagerEx.getInstanceEx(configuration.getProject());
+    final List<? extends BeforeRunTask> buildTasks = getBuildTasks(runManager, configuration);
+
+    if (hasBuildTaskForModule(buildTasks, module)) {
+      for (BeforeRunTask task : buildTasks) {
+        task.setEnabled(true);
+      }
+    } else {
+      BeforeRunTask buildTask = createBuildTask(module);
+      List<BeforeRunTask> tasks = runManager.getBeforeRunTasks(configuration);
+      tasks.add(buildTask);
+      runManager.setBeforeRunTasks(configuration, tasks, true);
+    }
+  }
+
+  @Override
```

Comment:
could you add some docs to explain what the difference is between these two overridden methods?